### PR TITLE
DL-14744: UTR Page

### PR DESCRIPTION
**Pull request by daglisood**
# JIRA Story or Github Number - descriptive title

**Bug fix / New feature** (delete as appropriate)

Please include a summary / description of the change and which issue it fixes.  Include any relevant user needs, context or links to other PRs related to this PR (eg. acceptance tests, environment config).

## Checklist

Reviewee (Replace with your name)

 - [ ]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [ ]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [ ]  I've executed the acceptance test pack locally to ensure there are no functional regressions
 - [ ]  I've added my code using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
 - [ ]  I've run a dependency check to ensure all dependencies are up to date

Reviewer (Replace with your name)

 - [ ]  I've confirmed that every effort has been made to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [ ]  I've confirmed appropriate tests has been included with any code added (Unit, Integration, Acceptance etc.)
 - [ ]  I've executed the acceptance test pack locally to ensure there are no functional regressions
 - [ ]  I've confirmed code was added using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
 - [ ]  I've run a dependency check to ensure all dependencies are up to date

### CODE REVIEW

**Review by christianmennerich — COMMENTED** [review]
A few general point we can discuss:

* Individuals with SA-CTR are not allowed to use the service, or (re)enrol, maybe we can reflect this in the auth method. I wonder if we could reflect this in the authorisation method.

* The added pages/controllers (for postcode, urn, utr) are for the new journey solely, and wired accordingly. Maybe we can reflect this by moving these to a dedicated package or reflect this in their names.

File: app/controllers/AwrsRegisteredPostcodeController.scala [hunk 1]
```diff
@@ -46,26 +44,30 @@ class AwrsRegisteredPostcodeController @Inject()(val mcc: MessagesControllerComp
...
-        }
-      else
-        Future.successful(NotFound)
+    enrollmentEligibleAuthorisedAction { implicit ar =>
+      restrictedAccessCheck {
+        if (awrsFeatureSwitches.enrolmentJourney().enabled)
+          keyStoreService.fetchAwrsRegisteredPostcode flatMap {
+            case Some(registeredPostcode) => Future.successful(Ok(template(awrsRegisteredPostcodeForm.form.fill(registeredPostcode))))
+            case _ => Future.successful(Ok(template(awrsRegisteredPostcodeForm.form)))
+          }
```
Comment (CEPoole) [inline app/controllers/AwrsRegisteredPostcodeController.scala:53]: Having both `flatMap` and the `Future.successful`s here is redundant. A simple map will remove the need for futures.

File: app/forms/AwrsEnrolmentUtrForm.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      val actualUtr = utr.trim.takeRight(10).toList
+      val checkDigit = actualUtr.head.asDigit
+      val restOfUtr = actualUtr.tail
+      val weights = List(six, seven, eight, nine, ten, five, four, three, two)
+      val weightedUtr = for ((w1, u1) <- weights zip restOfUtr) yield w1 * u1.asDigit
+      val total = weightedUtr.sum
+      val remainder = total % eleven
+      isValidUtr(remainder, checkDigit)
+    }
+  }
```
Comment (CEPoole) [inline app/forms/AwrsEnrolmentUtrForm.scala:47]: Might be better placed in a utils file rather than tied to a particular form? This could then be used to validate UTRs in other parts of the service if needed.

File: app/services/BusinessMatchingService.scala [hunk 3]
```diff
@@ -69,6 +72,30 @@ class BusinessMatchingService @Inject()(keyStoreService: KeyStoreService,
     }
   }
 
+  def isValidUTRandPostCode(utr: String, postCode: AwrsRegisteredPostcode, authRetrievals: StandardAuthRetrievals, isSA: Boolean)
+                           (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] = {
+    val searchData = MatchBusinessData(acknowledgementReference = SessionUtil.getUniqueAckNo,
+      utr = utr, individual = None, organisation = None)
```
Comment (CEPoole) [inline app/services/BusinessMatchingService.scala:75]: suspicious shadowing by having two `utr`s in scope. If we were to rename one of these, the code will not complain despite the behaviour changing.

File: app/views/awrs_urn.scala.html [hunk 4]
```diff
@@ -44,7 +44,7 @@
     }
 
     <header class="page-header">
-        <h1 id="businessNameChangeConfirmation-heading" class="govuk-heading-xl">@messages(headerTitleName)</h1>
+        <h1 id="awrsUrn-heading" class="govuk-heading-xl">@messages(headerTitleName)</h1>
```
Comment (CEPoole) [inline app/views/awrs_urn.scala.html:47]: any risk that changing this ID breaks an acceptance test?

File: conf/app.routes [hunk 5]
```diff
@@ -26,6 +26,8 @@ GET         /unauthorised                                   controllers.IndexCon
 
 GET         /urn                                            controllers.AwrsUrnController.showArwsUrnPage
 POST        /urn/saveAndReturn                              controllers.AwrsUrnController.saveAndContinue
+GET         /utr                                            controllers.AwrsUtrController.showArwsUtrPage
+POST        /utr/saveAndReturn                              controllers.AwrsUtrController.saveAndContinue
```
Comment (CEPoole) [inline conf/app.routes:30]: why have these been duplicated?

File: app/services/BusinessMatchingService.scala [hunk 6]
```diff
@@ -69,6 +72,30 @@ class BusinessMatchingService @Inject()(keyStoreService: KeyStoreService,
...
+    businessMatchingConnector.lookup(searchData, if (isSA) "sa" else "org", authRetrievals) map {
+      validatePostCode(postCode.registeredPostcode, _)
+    }
+  }
+
+
+  private def validatePostCode(postcode: String, dataReturned: JsValue): Boolean = {
```
Comment (CEPoole) [inline app/services/BusinessMatchingService.scala:82]: We should probably call this `verify` rather than `validate`, as this function only checks that they match. If the Postcodes were both email addresses it would still succeed.
The frontend forms do use `addressMapping` to check the Postcode regex, but there's an assumption in this function that it has been done.

File: test/services/BusinessMatchingServiceTest.scala [hunk 7]
```diff
@@ -36,6 +36,24 @@ class BusinessMatchingServiceTest extends AwrsUnitTestTraits with MockKeyStoreSe
   val mockBusinessMatchingConnector: BusinessMatchingConnector = mock[BusinessMatchingConnector]
   val businessMatchingServiceTest: BusinessMatchingService = new BusinessMatchingService(testKeyStoreService, mockBusinessMatchingConnector, testSave4LaterService, mockAuditable)
 
+  "isValidUTRandPostCode" must {
+    "validate a UTR and PostCode is match for an individual" in {
+      when(mockBusinessMatchingConnector.lookup(ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())(ArgumentMatchers.any(), ArgumentMatchers.any())).thenReturn(Future.successful(matchSuccessResponseJsonWithPostCode("WD6 5DR")))
+      val result = businessMatchingServiceTest.isValidUTRandPostCode(testUtr, AwrsRegisteredPostcode("WD65dr"), TestUtil.defaultAuthRetrieval, true)
```
Comment (CEPoole) [inline test/services/BusinessMatchingServiceTest.scala:42]: No test for an invalid Postcode in a `validate` function. See my other comment.

File: app/connectors/TaxEnrolmentsConnector.scala [hunk 8]
```diff
@@ -61,6 +61,17 @@ class TaxEnrolmentsConnector @Inject()(servicesConfig: ServicesConfig,
     response
   }
 
+  def enrol(requestPayload: RequestPayload,
```
[the remark this answers is not in the document]
Comment (CEPoole) [reply app/connectors/TaxEnrolmentsConnector.scala:64]: Code is still commented and should be removed

File: app/services/BusinessMatchingService.scala (hunk 6 above)
Comment (CEPoole) [reply app/services/BusinessMatchingService.scala:82]: eg. Added this test and it passed
```scala
    "validate a UTR and PostCode is match for an email" in {
      when(mockBusinessMatchingConnector.lookup(ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())(ArgumentMatchers.any(), ArgumentMatchers.any())).thenReturn(Future.successful(matchSuccessResponseJsonWithPostCode("omg@gmail.com")))
      val result = businessMatchingServiceTest.isValidUTRandPostCode(testUtr, AwrsRegisteredPostcode("omg@gmail.com"), TestUtil.defaultAuthRetrieval, true)
      await(result) mustBe true
    }

**Review by CEPoole — CHANGES_REQUESTED** [review]

File: app/services/EnrolService.scala [hunk 9]
```diff
@@ -54,6 +53,25 @@ class EnrolService @Inject()(taxEnrolmentsConnector: TaxEnrolmentsConnector,
       }
     }
 
+  def enrolAWRS(awrsRef: String,
```
[the remark this answers is not in the document]
Comment (daglisood) [reply app/services/EnrolService.scala:56]: done

File: app/controllers/auth/AuthFunctionality.scala [hunk 10]
```diff
@@ -66,6 +66,7 @@ trait AuthFunctionality extends AuthorisedFunctions with Logging {
       } recover recoverAuthorisedCalls
```
[the remark this answers is not in the document]
Comment (daglisood) [reply app/controllers/auth/AuthFunctionality.scala:66]: done

File: app/connectors/TaxEnrolmentsConnector.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (daglisood) [reply app/connectors/TaxEnrolmentsConnector.scala:64]: Audit map has been moved to a parameter so we can reuse one enrol method

File: test/forms/AwrsEnrolmentUtrFormSpec.scala [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import forms.AwrsEnrolmentUrnForm.maxQueryLength
+import forms.AwrsEnrolmentUtrForm.{awrsEnrolmentUtrForm, utr}
+import forms.test.util.AwrsFormTestUtils
+import models.AwrsEnrolmentUtr
+import org.scalatest.BeforeAndAfterEach
+import org.scalatestplus.mockito.MockitoSugar
+import org.scalatestplus.play.PlaySpec
+import play.api.data.Form
+
+class AwrsEnrolmentUtrFormSpec extends PlaySpec with MockitoSugar with BeforeAndAfterEach with AwrsFormTestUtils {
```
[the remark this answers is not in the document]
Comment (daglisood) [reply test/forms/AwrsEnrolmentUtrFormSpec.scala:28]: removed

File: app/forms/AwrsEnrolmentUtrForm.scala (hunk 2 above)
Comment (daglisood) [reply app/forms/AwrsEnrolmentUtrForm.scala:47]: done

File: app/controllers/AwrsRegisteredPostcodeController.scala (hunk 1 above)
Comment (daglisood) [reply app/controllers/AwrsRegisteredPostcodeController.scala:53]: done

File: app/views/awrs_urn.scala.html (hunk 4 above)
Comment (daglisood) [reply app/views/awrs_urn.scala.html:47]: no acceptance tests for this journey yet

File: app/services/BusinessMatchingService.scala (hunk 6 above)
Comment (daglisood) [reply app/services/BusinessMatchingService.scala:82]: I have renamed the function. Though i do not thin k we should (re) validate postcode or utr here again. Both the UTR and the Postcode have been validated in their respective forms already and at this stage of the journey they are deemed to be in correct format.

File: test/services/BusinessMatchingServiceTest.scala (hunk 7 above)
Comment (daglisood) [reply test/services/BusinessMatchingServiceTest.scala:42]: I have renamed the function to not be validate anymore. Was misleading. We are only matching postcode/utr to what we receive from business matching service

File: app/services/BusinessMatchingService.scala (hunk 6 above)
Comment (CEPoole) [reply app/services/BusinessMatchingService.scala]: I agree

File: app/controllers/AwrsUtrController.scala [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+          awrsEnrolmentUtrForm.bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(template(formWithErrors))),
+            utr => {
+              keyStoreService.saveAwrsEnrolmentUtr(utr)
+              keyStoreService.fetchAwrsUrnSearchResult.flatMap {sr =>
+                keyStoreService.fetchAwrsRegisteredPostcode.flatMap { pc =>
+                  businessMatchingService.verifyUTRandPostCode(utr.utr, pc.get, ar, isSA).flatMap { utrPostCodeMatch: Boolean =>
+                    if(utrPostCodeMatch) {
+                      enrolService.enrolAWRS(sr.get.results.head.awrsRef,
+                        pc.get.registeredPostcode, Some(utr.utr),
```
Comment (CEPoole) [inline app/controllers/AwrsUtrController.scala:76]: Both of these lines have unsafe `.get`s, and there are no tests for this.
Adding tests shows that if either of these are `None` then we get unhandled exceptions

File: app/controllers/AwrsUtrController.scala [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+          keyStoreService.fetchAwrsEnrolmentUtr flatMap {
+            case Some(utr) => Future.successful(Ok(template(awrsEnrolmentUtrForm.form.fill(utr), isSA)))
+            case _ => Future.successful(Ok(template(awrsEnrolmentUtrForm.form, isSA)))
+          }
+        } else Future.successful(NotFound)
+      }
+    }
+  }
+
+  private def safeGet[T](x: Option[T]): T = x.fold(throw new RuntimeException(s"No value found for ${x.getClass.getName} fron keystore"))(identity)
```
Comment (christianmennerich) [inline app/controllers/AwrsUtrController.scala:62]: Typo in the error message.

I am wondering why we can get here with `None`, as we always expect a value to be present - and if a more specific error  message is possible?

File: app/controllers/AwrsUtrController.scala [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      restrictedAccessCheck {
+        if (awrsFeatureSwitches.enrolmentJourney().enabled) {
+          val isSA = accountUtils.isSaAccount(ar.enrolments).getOrElse(false)
+          awrsEnrolmentUtrForm.bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(template(formWithErrors))),
+            utr => {
+              keyStoreService.saveAwrsEnrolmentUtr(utr)
+              keyStoreService.fetchAwrsUrnSearchResult.flatMap {sr =>
+                keyStoreService.fetchAwrsRegisteredPostcode.flatMap { pc =>
+                  businessMatchingService.verifyUTRandPostCode(utr.utr, pc.get, ar, isSA).flatMap { utrPostCodeMatch: Boolean =>
```
Comment (christianmennerich) [inline app/controllers/AwrsUtrController.scala:75]: Another unsafe get?

**Review by christianmennerich — CHANGES_REQUESTED** [review]

File: app/connectors/TaxEnrolmentsConnector.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (christianmennerich) [reply app/connectors/TaxEnrolmentsConnector.scala]: `enrol` in the `TaxEnrolmentsConnector` is only called one time atm (from `enrolAWRS` in the `EnrolService`). So, default empty Map is not used. But we could have the auditMap default to empty map in `enrolAWRS`, and then we would not need to hand in an empty map when calling `enrolAWRS` in from the `AwrsUtrController`? Depending on where we want to make it apparent where we want to activate audit logging.

File: test/forms/AwrsEnrolmentUtrFormSpec.scala [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package forms
+
+import forms.AwrsEnrolmentUrnForm.maxQueryLength
```
[the remark this answers is not in the document]
Comment (christianmennerich) [reply test/forms/AwrsEnrolmentUtrFormSpec.scala]: I still do not get the "field is more than max length" test to fail (but by handing in empty input)...

File: app/forms/validation/util/UTRValidator.scala [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+      val actualUtr = utr.trim.takeRight(10).toList
+      val checkDigit = actualUtr.head.asDigit
+      val restOfUtr = actualUtr.tail
+      val weights = List(six, seven, eight, nine, ten, five, four, three, two)
+      val weightedUtr = for ((w1, u1) <- weights zip restOfUtr) yield w1 * u1.asDigit
+      val total = weightedUtr.sum
+      val remainder = total % eleven
+      isValidUtr(remainder, checkDigit)
+    }
+  }
```
Comment (christianmennerich) [inline app/forms/validation/util/UTRValidator.scala:40]: As this validation is now an object by its own, I am wondering if we could pull in more logic here we are doing at the moment in the calling `AwrsEnrolmentUtrForm` (besides trimming we do in both places replacement of blanks that is at the moment only done as part of the form validation (not in the UTRValidator). Would need a bit more refactoring, as we would need to return validation failure reason from here, so we can think about this later?

Either way, some more tests would be good (ensuring UTRs are still recognised when formatted containing strings).

File: test/forms/AwrsEnrolmentUtrFormTest.scala [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+      form.bind(Map(fieldId -> "a" * 141)).fold(
+        formWithErrors => {
+          formWithErrors(fieldId).errors.size mustBe 1
+          messages(formWithErrors(fieldId).errors.head.message) mustBe messages("awrs.utr.length", fieldNameInErrorMessage)
+        },
+        _ => fail("Field should contain errors")
+      )
+    }
+
+    "field is more than length mismatch" in {
```
Comment (christianmennerich) [inline test/forms/AwrsEnrolmentUtrFormTest.scala:55]: This test does not violate "awrs.utr.length" validation, but passed UTR is invalid. So, entered number should be longer, and expectation changed to "awrs.utr.length" to match indicated test.

**Review by christianmennerich — DISMISSED** [review]

File: app/forms/validation/util/UTRValidator.scala (hunk 16 above)
Comment (daglisood) [reply app/forms/validation/util/UTRValidator.scala:40]: I'll add this to the tidy up task. but i do agree with the suggestion

File: test/forms/AwrsEnrolmentUtrFormTest.scala (hunk 17 above)
Comment (daglisood) [reply test/forms/AwrsEnrolmentUtrFormTest.scala:55]: I have removed this test. For UTR we have to test for the string to be 10 or 13 char long

Comment (christianmennerich) [reply test/forms/AwrsEnrolmentUtrFormTest.scala:44]: I still think we can test for the accepted 'formattings' here (i.e. the number string containing blanks at any place - start, end, in-between), as deleting these is part of the preparation in the form.

File: app/controllers/AwrsUtrController.scala [hunk 18]
```diff
@@ -72,12 +72,11 @@ class AwrsUtrController @Inject()(mcc: MessagesControllerComponents,
               keyStoreService.saveAwrsEnrolmentUtr(utr)
               keyStoreService.fetchAwrsUrnSearchResult.flatMap {sr =>
                 keyStoreService.fetchAwrsRegisteredPostcode.flatMap { pc =>
-                  businessMatchingService.verifyUTRandPostCode(utr.utr, pc.get, ar, isSA).flatMap { utrPostCodeMatch: Boolean =>
+                  businessMatchingService.verifyUTRandPostCode(utr.utr, safeGet(pc), ar, isSA).flatMap { utrPostCodeMatch: Boolean =>
```
Comment (CEPoole) [inline app/controllers/AwrsUtrController.scala:75]: There is no difference in behaviour between `get` and `safeGet` besides adding "keystore" to the error.
This still leaves the user stuck and unable to continue. Has anyone asked the designers what they would want the user to see in this situation to prevent contact?

Comment (daglisood) [reply app/controllers/AwrsUtrController.scala:75]: This is read from key store. We can't default it to anything and the only reason it will fail is if someone had bookmarked the pages and we may get None from keystore. I think thowing an exception is a good recourse

Comment (CEPoole) [reply app/controllers/AwrsUtrController.scala:75]: But the user can fix this easily by re-entering a Postcode on the previous page. I'm almost certain that a designer would prefer that we redirect to that page, or show a deliberate error page saying something like "Your details could not be found, restart your journey to try again"?
